internal/api/poller.go

```diff
@@ -36,56 +40,138 @@ func (a SingleRunPoller) Run(ctx context.Context, ch <-chan struct{}, action Pol
 
 // InterruptiblePoller is a poller than runs an action and allows for interrupts.
 type InterruptiblePoller struct {
+	name string
 }
 
 var _ Poller = (*InterruptiblePoller)(nil)
 
 // NewInterruptiblePoller creates an instance of InterruptiblePoller.
-func NewInterruptiblePoller() *InterruptiblePoller {
-	return &InterruptiblePoller{}
+func NewInterruptiblePoller(name string) *InterruptiblePoller {
+	return &InterruptiblePoller{name: name}
 }
 
 // Run runs the poller.
 func (ip *InterruptiblePoller) Run(ctx context.Context, ch <-chan struct{}, action PollerFunc, resetDuration time.Duration) {
-	timer := time.NewTimer(0)
-	var cancel context.CancelFunc
-	var mu sync.Mutex
+	logger := log.From(ctx).With("poller-name", ip.name)
+	if ip.name != "content" {
+		logger = log.NopLogger()
+	}
+	ctx = log.WithLoggerContext(ctx, logger)
 
-	go func() {
-		for _ = range ch {
-			mu.Lock()
-			if cancel != nil {
-				cancel()
+	jt := initJobTracker(ctx, action)
+	defer jt.clear()
+
+	pollerQueue := make(chan job, 100)
+
+	worker := func() {
+		for j := range pollerQueue {
+			select {
+			case <-j.ctx.Done():
+			case <-j.run():
+				if j.ctx.Err() == nil {
+					<-time.After(resetDuration)
+					pollerQueue <- jt.create()
+				}
 			}
-			mu.Unlock()
+		}
+	}
+
+	for i := 0; i < 2; i++ {
+		go worker()
+	}
 
-			timer.Reset(0)
+	go func() {
+		for range ch {
+			// cancel all existing jobs
+			jt.clear()
+			pollerQueue <- jt.create()
 		}
 	}()
 
-	done := false
-	for !done {
+	pollerQueue <- jt.create()
+
+	<-ctx.Done()
+}
+
+type job struct {
+	id         uuid.UUID
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+	action     PollerFunc
+}
+
+func (j *job) run() <-chan bool {
+	ch := make(chan bool, 1)
+
+	done := make(chan bool, 1)
+
+	go func() {
+		j.action(j.ctx)
+		done <- true
+	}()
+
+	go func() {
 		select {
-		case <-ctx.Done():
-			done = true
-			break
-		case <-timer.C:
-			var actionContext context.Context
-			go func() {
-				mu.Lock()
-				actionContext, cancel = context.WithCancel(ctx)
-				mu.Unlock()
-
-				rerun := action(actionContext)
-				if actionContext.Err() == nil {
-					dur := resetDuration
-					if rerun {
-						dur = 0
-					}
-					timer.Reset(dur)
-				}
-				cancel()
-			}()
+		case <-j.ctx.Done():
```

Comment:
do we want to reduce these to a single case?

Reply:
We need both cases. This particular case means the context for the job was canceled. Nothing else should happen here. The second case will fire if the job was ran.